ext/liquid_c/resource_limits.c

```diff
@@ -0,0 +1,282 @@
+
+    return LONG2NUM(resource_limits->assign_score_limit);
+}
+
+static VALUE resource_limits_set_assign_score_limit_method(VALUE self, VALUE assign_score_limit)
+{
+    resource_limits_t *resource_limits;
+    ResourceLimits_Get_Struct(self, resource_limits);
+
+    if (assign_score_limit == Qnil) {
+        resource_limits->assign_score_limit = LONG_MAX;
+    } else {
+        resource_limits->assign_score_limit = NUM2LONG(assign_score_limit);
+    }
+
+    return Qnil;
+}
+
+static VALUE resource_limits_render_score_method(VALUE self)
+{
+    resource_limits_t *resource_limits;
+    ResourceLimits_Get_Struct(self, resource_limits);
+
+    return LONG2NUM(resource_limits->render_score);
+}
+
+static VALUE resource_limits_assign_score_method(VALUE self)
+{
+    resource_limits_t *resource_limits;
+    ResourceLimits_Get_Struct(self, resource_limits);
+
+    return LONG2NUM(resource_limits->assign_score);
+}
+
+static VALUE resource_limits_initialize_method(VALUE self, VALUE render_length_limit,
+                                               VALUE render_score_limit, VALUE assign_score_limit)
+{
+    resource_limits_set_render_length_limit_method(self, render_length_limit);
+    resource_limits_set_render_score_limit_method(self, render_score_limit);
+    resource_limits_set_assign_score_limit_method(self, assign_score_limit);
+
+    return Qnil;
+}
+
+__attribute__((noreturn))
+void resource_limits_raise_limits_reached(resource_limits_t *resource_limit)
+{
+    resource_limit->reached_limit = true;
+    rb_raise(cMemoryError, "Memory limits exceeded");
+}
+
+void resource_limits_increment_render_score(resource_limits_t *resource_limits, long amount)
+{
+    resource_limits->render_score = resource_limits->render_score + amount;
+
+    if (resource_limits->render_score > resource_limits->render_score_limit) {
+        resource_limits_raise_limits_reached(resource_limits);
+    }
+}
+
+static VALUE resource_limits_increment_render_score_method(VALUE self, VALUE amount)
+{
+    resource_limits_t *resource_limits;
+    ResourceLimits_Get_Struct(self, resource_limits);
+
+    resource_limits_increment_render_score(resource_limits, NUM2LONG(amount));
+
+    return Qnil;
+}
+
+static void resource_limits_increment_assign_score(resource_limits_t *resource_limits, long amount)
+{
+    resource_limits->assign_score = resource_limits->assign_score + amount;
+
+    if (resource_limits->assign_score > resource_limits->assign_score_limit) {
+        resource_limits_raise_limits_reached(resource_limits);
+    }
+}
+
+static VALUE resource_limits_increment_assign_score_method(VALUE self, VALUE amount)
+{
+    resource_limits_t *resource_limits;
+    ResourceLimits_Get_Struct(self, resource_limits);
+
+    resource_limits_increment_assign_score(resource_limits, NUM2LONG(amount));
+
+    return Qnil;
+}
+
+void resource_limits_increment_write_score(resource_limits_t *resource_limits, VALUE output)
+{
+    Check_Type(output, T_STRING);
```

Comment:
In https://github.com/Shopify/liquid-c/pull/64#discussion_r498825055 I meant that this should be done in resource_limits_increment_write_score_method to avoid doing it redundantly in the VM loop.